CRM/Contact/BAO/Query.php

```diff
@@ -6136,62 +6161,91 @@ protected function prepareOrderBy($sort, $sortByChar, $sortOrder, $additionalFro
         }
       }
       elseif ($sortByChar) {
-        $order = " ORDER BY UPPER(LEFT(contact_a.sort_name, 1)) asc";
+        $orderByArray = array("UPPER(LEFT(contact_a.sort_name, 1)) asc");
       }
       else {
         $order = " ORDER BY contact_a.sort_name asc, contact_a.id";
       }
     }
+    if (!$order && empty($orderByArray)) {
```

Comment:
Refactoring in this function ensures an array of order-by fields although it is still pretty funky. Previously the function did not iterate through all fields - just the first one in the string, to add special handling.

The handling we are adding is a way of avoiding joining the option_value table in when sorting by it. By rendering the values in php & using field($fieldName, 4,6,7) I was able to render the query without the join. Results in test were 10 times faster, in some cases much more.